pkg/operator/controller/ingress/controller.go

```diff
@@ -179,18 +179,19 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		return reconcile.Result{}, fmt.Errorf("failed to get network 'cluster': %v", err)
 	}
 
-	// Admit if necessary. Don't process until admission succeeds. If admission is
-	// successful, immediately re-queue to refresh state.
-	if !isAdmitted(ingress) || needsReadmission(ingress) {
-		if err := r.admit(ingress, ingressConfig, infraConfig); err != nil {
-			switch err := err.(type) {
-			case *admissionRejection:
-				r.recorder.Event(ingress, "Warning", "Rejected", err.Reason)
-				return reconcile.Result{}, nil
-			default:
-				return reconcile.Result{}, fmt.Errorf("failed to admit ingresscontroller: %v", err)
-			}
+	// Admit the ingresscontroller. Don't process until admission succeeds. If admission results
+	// in an update, immediately re-queue to refresh state.
+	err, updated := r.admit(ingress, ingressConfig, infraConfig)
```

Comment:
@Miciah 

After an upgrade, existing admitted ingresscontrollers will have a nil `.spec.routeAdmission` and _should_ have a new generation due to the container image update, triggering re-admission. We have an internal invariant that `.status` will always be populated with defaults, which happens during admission. Consequently, if we ever fail to re-admit upgraded ingresscontrollers for any reason and then process the previously admitted state, we run the risk of violating our invariant and can panic.

Rather than bank on the generation changing, I decided to be even more defensive and start unconditionally admitting ingresscontrollers during reconciliation. This is more expensive in terms of CPU (we will hit our caches for list calls and do some in-memory processing), but the most part, the call should no a persistence no-op, and makes me feel more confident we'll always get things into a valid state before processing.

Is this too paranoid?

Reply:
I think the CPU expense is worth preventing the possibility of a panic.

Reply:
Oops... this actually doesn't work because `admit()` won't mutate the input, so defaults aren't set and visible to the caller unless (as happens today on master) a persisted update occurs, a re-queue happens, and the state is refreshed during the next reconcile...

This might need re-thought...

Reply:
Oops again — the problem is that status _should_ have been persisted. I needed to update `ingressStatusesEqual()` to incorporate the route admission status field, causing a diff and ultimately persistence of status value.

Reply:
Nevermind, as of https://github.com/openshift/api/pull/565 we won't be putting the field on status anyway.